tensorflow_addons/layers/spatial_pyramid_pooling.py

```diff
@@ -0,0 +1,123 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Spatial Pyramid Pooling layers"""
+
+import tensorflow as tf
+from tensorflow_addons.layers.adaptive_pooling import AdaptiveAveragePooling2D
+import tensorflow_addons.utils.keras_utils as conv_utils
+
+from typeguard import typechecked
+from typing import Union, Iterable
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class SpatialPyramidPooling2D(tf.keras.layers.Layer):
+    """Performs Spatial Pyramid Pooling.
+
+    Original Paper: https://arxiv.org/pdf/1406.4729.pdf
+
+    Spatial Pyramid Pooling generates a fixed-length representation
+    regardless of input size/scale. It is typically used before layers
+    that require a constant input shape, for example before Dense Layers.
+
+    Arguments:
+      bins: Either a collection of integers or a collection of collections of 2 integers.
+        Each element in the inner collection must contain 2 integers, (pooled_rows, pooled_cols)
+      data_format: A string,
+        one of `channels_last` (default) or `channels_first`.
+        The ordering of the dimensions in the inputs.
+        `channels_last` corresponds to inputs with shape
+        `(batch, height, width, channels)` while `channels_first`
+        corresponds to inputs with shape `(batch, channels, height, width)`.
+
+    Input shape:
+      - If `data_format='channels_last'`:
+        4D tensor with shape `(batch_size, height, width, channels)`.
+      - If `data_format='channels_first'`:
+        4D tensor with shape `(batch_size, channels, height, width)`.
+
+    Output shape:
+      The output is the pooled image, flattened across its height and width
+      - If `data_format='channels_last'`:
+        3D tensor with shape `(batch_size, num_bins, channels)`.
+      - If `data_format='channels_first'`:
+        3D tensor with shape `(batch_size, channels, num_bins)`.
+    """
+
+    @typechecked
+    def __init__(
+        self,
+        bins: Union[Iterable[int], Iterable[Iterable[int]]],
+        data_format=None,
+        *args,
+        **kwargs
+    ):
+        self.bins = [conv_utils.normalize_tuple(bin, 2, "bin") for bin in bins]
+        self.data_format = conv_utils.normalize_data_format(data_format)
+        self.pool_layers = []
+        for bin in self.bins:
+            self.pool_layers.append(AdaptiveAveragePooling2D(bin, self.data_format))
+        super().__init__(*args, **kwargs)
+
+    def call(self, inputs, **kwargs):
+        dynamic_input_shape = tf.shape(inputs)
+        outputs = []
+        index = 0
+        if self.data_format == "channels_last":
+            for bin in self.bins:
+                new_inp = inputs[
```

Comment:
The slices are not well readable. Can you refactor this to make it please?

Reply:
Do you mean using `tf.slice` or simply expanding the internal terms? TensorFlow documentation recommends using the pythonic way.

Reply:
No, I mean the way the code is laid out, it's less readable.